src/io.rs

```diff
@@ -719,6 +721,16 @@ where
 
     /// Handles a new item to send out that arrived through the incoming channel.
     fn handle_incoming_item(&mut self, item: QueuedItem) -> Result<(), LocalProtocolViolation> {
+        // Process the wait queue to avoid this new item "jumping the queue".
+        match &item {
```

Comment:
https://github.com/casper-network/juliet/pull/5/commits/1dff6627c516247286f51910decbf6e3615db54e seems to add starvation protection, i.e. newer data cannot consistently get in front of existing data. Was this behavior observed to be problematic?

My core issue with this is that if we process the wait queue each time anyway, it might be better to not even check if we can bypass it and just put everything in the wait queue every time. However, processing the wait queue is expensive, especially if the previously mentioned change is made. Queuing messages will then result in quadratic complexity!

Reply:
It was problematic in that Alice's requests started timing out, as ones in the wait queue weren't processed since newer ones kept getting preferential treatment.

I did consider just dumping everything in the wait queue, however I had the same reservation as you about the cost (and it also seemed to be somewhat abusing the intent of the wait queue - it would at least need renamed for clarity I think if we did that).

Reply:
> At the very least add `QueuedItem::is_request` :)

I can do, but tbh I don't see why we'd want that or where we'd use it?